### PR TITLE
Show running tests

### LINKED ISSUE #3728: Show currently running tests in progress output
## Summary

Part of #2162 that was not implemented, because we want to compare the experience between single dll run, and multi dll run (in dotnet test), once we have it available.

By default please also output currently running test + result of passing tests. This is also important when a test does actually hang/crash so you can see what test might have caused it.

## Background and Motivation

Show which tests are running so we know when a test is hanging.

## Proposed Feature

Show one line for every test in progress and timing. 

Make sure the information fits on screen. Decide which tests should be cut off when there are too many to still keep the output useful.

## Alternative Designs

Consider the impact of this https://github.com/microsoft/testfx/issues/3639 where I think it is more appropriate to implement detection of stuck tests closer to the test framework. This also limits the amount of information we will render on screen, and we don't have to consier what will happen when tests run very fast, or how will user figure out if test is stuck.

### PR DESCRIPTION
**Pull request by drognanar**
Fixes #3728

Reports the progress of actively running tests:

image

### CODE REVIEW

**Comment by Evangelink** [comment]
Looking dope!

progress-1000-tests

I do see a few flickering effect depending on machine load.

**Comment by nohwnd** [comment]
True. My only objection is the “and 1 more”, just show the test that is running :)

-j
________________________________
Od: Amaury Levé ***@***.***>
Odesláno: Wednesday, December 4, 2024 9:08:48 PM
Komu: microsoft/testfx ***@***.***>
Kopie: Subscribed ***@***.***>
Předmět: Re: [microsoft/testfx] Show running tests (PR #4221)


Looking dope!

progress-1000-tests.gif (view on web)<https://github.com/user-attachments/assets/d7ea6295-49b5-4487-918d-6a59509193b8>

I do see a few flickering effect depending on machine load.

**Comment by Evangelink** [comment]
> My only objection is the “and 1 more”, just show the test that is running :)

Yes!

File: src/Platform/Microsoft.Testing.Platform/OutputDevice/Terminal/AnsiTerminalTestProgressFrame.cs [hunk 1]
```diff
@@ -228,12 +296,57 @@ public void Render(AnsiTerminalTestProgressFrame previousFrame, AnsiTerminal ter
...
 
-    public void Clear() => ProgressCount = 0;
+    private List<object> GenerateLinesToRender(TestProgressState?[] progress)
+    {
+        var linesToRender = new List<object>(progress.Length);
+
+        // Note: We want to render the list of active tests, but this can easily fill up the full screen.
+        // As such, we should balance the number of active tests shown per project.
+        // We do this by distributing the remaining lines for each projects.
+        var progressItems = progress.OfType<TestProgressState>().ToList();
```
Comment (Evangelink) [inline src/Platform/Microsoft.Testing.Platform/OutputDevice/Terminal/AnsiTerminalTestProgressFrame.cs]: Nit: ToArray is probably sufficient here.

File: src/Platform/Microsoft.Testing.Platform/OutputDevice/Terminal/AnsiTerminalTestProgressFrame.cs [hunk 2]
```diff
@@ -228,12 +296,57 @@ public void Render(AnsiTerminalTestProgressFrame previousFrame, AnsiTerminal ter
...
+        int linesToDistribute = (int)(Height * 0.7) - 1 - progressItems.Count;
+        var detailItems = new IEnumerable<TestDetailState>[progressItems.Count];
+        var sortedItemsI = Enumerable.Range(0, progressItems.Count).OrderBy(i => progressItems[i].TestNodeResultsState?.Count ?? 0).ToList();
```
Comment (Evangelink) [inline src/Platform/Microsoft.Testing.Platform/OutputDevice/Terminal/AnsiTerminalTestProgressFrame.cs]: We don't need the `ToList` as we are going over the items only once

File: src/Platform/Microsoft.Testing.Platform/OutputDevice/Terminal/TerminalTestReporter.cs [hunk 3]
```diff
@@ -108,10 +108,12 @@ private static Regex GetFrameRegex()
     }
 #endif
 
+    private int _counter;
+
     /// <summary>
     /// Initializes a new instance of the <see cref="TerminalTestReporter"/> class with custom terminal and manual refresh for testing.
     /// </summary>
-    internal TerminalTestReporter(IConsole console, TerminalTestReporterOptions options)
+    public TerminalTestReporter(IConsole console, TerminalTestReporterOptions options)
```
Comment (nohwnd) [inline src/Platform/Microsoft.Testing.Platform/OutputDevice/Terminal/TerminalTestReporter.cs:128]: Please add option for showing test details so we can disable this in `dotnet test` until it is ready for multi project runs.

File: src/Platform/Microsoft.Testing.Platform/OutputDevice/Terminal/AnsiTerminalTestProgressFrame.cs (hunk 2 above)
Comment (Evangelink) [inline src/Platform/Microsoft.Testing.Platform/OutputDevice/Terminal/AnsiTerminalTestProgressFrame.cs]: What do you mean by the `I` suffix?

File: src/Platform/Microsoft.Testing.Platform/OutputDevice/Terminal/AnsiTerminalTestProgressFrame.cs [hunk 4]
```diff
@@ -228,12 +296,57 @@ public void Render(AnsiTerminalTestProgressFrame previousFrame, AnsiTerminal ter
...
+            linesToRender.Add(progressItems[progressI]);
+            linesToRender.AddRange(detailItems[progressI]);
+        }
+
+        return linesToRender;
+    }
+
+    public void Clear() => RenderedLines?.Clear();
+
+    internal class RenderedProgressItem
```
Comment (Evangelink) [inline src/Platform/Microsoft.Testing.Platform/OutputDevice/Terminal/AnsiTerminalTestProgressFrame.cs]: Seal it

File: src/Platform/Microsoft.Testing.Platform/OutputDevice/Terminal/AnsiTerminalTestProgressFrame.cs (hunk 2 above)
Comment (drognanar) [reply src/Platform/Microsoft.Testing.Platform/OutputDevice/Terminal/AnsiTerminalTestProgressFrame.cs]: that the list contains the indices of sorted items. could have used `Indices` instead

Comment (Evangelink) [reply src/Platform/Microsoft.Testing.Platform/OutputDevice/Terminal/AnsiTerminalTestProgressFrame.cs]: Oh ok yeah please do, let's avoid abbreviations.

**Comment by Evangelink** [comment]
I think it'd be nice to also update the message `... and X more` when it's the only line being displayed
image

File: src/Platform/Microsoft.Testing.Platform/OutputDevice/Terminal/TerminalTestReporter.cs (hunk 3 above)
Comment (drognanar) [reply src/Platform/Microsoft.Testing.Platform/OutputDevice/Terminal/TerminalTestReporter.cs:128]: added the option to TerminalTestReporterOptions

File: src/Platform/Microsoft.Testing.Platform/OutputDevice/Terminal/AnsiTerminalTestProgressFrame.cs (hunk 2 above)
Comment (drognanar) [reply src/Platform/Microsoft.Testing.Platform/OutputDevice/Terminal/AnsiTerminalTestProgressFrame.cs]: changed

File: src/Platform/Microsoft.Testing.Platform/OutputDevice/Terminal/AnsiTerminalTestProgressFrame.cs (hunk 4 above)
Comment (drognanar) [reply src/Platform/Microsoft.Testing.Platform/OutputDevice/Terminal/AnsiTerminalTestProgressFrame.cs]: changed

File: src/Platform/Microsoft.Testing.Platform/OutputDevice/Terminal/AnsiTerminalTestProgressFrame.cs (hunk 2 above)
Comment (drognanar) [reply src/Platform/Microsoft.Testing.Platform/OutputDevice/Terminal/AnsiTerminalTestProgressFrame.cs]: changed

File: src/Platform/Microsoft.Testing.Platform/OutputDevice/Terminal/AnsiTerminalTestProgressFrame.cs (hunk 1 above)
Comment (drognanar) [reply src/Platform/Microsoft.Testing.Platform/OutputDevice/Terminal/AnsiTerminalTestProgressFrame.cs]: changed

**Comment by drognanar** [comment]
Specialized the message when no active tests can be shown on the terminal

**Review by Evangelink — DISMISSED** [review]
Functionnaly looking good. @nohwnd I'll let you finish the technical review.

@drognanar How hard would it be to add a few test in `TerminalTestReporterTests.cs` for the progress?

File: src/Platform/Microsoft.Testing.Platform/OutputDevice/Terminal/TerminalTestReporter.cs [hunk 5]
```diff
@@ -980,4 +990,24 @@ private static TerminalColor ToTerminalColor(ConsoleColor consoleColor)
...
+        string displayName,
+        string? executionId)
+    {
+        TestProgressState asm = _assemblies[$"{assembly}|{targetFramework}|{architecture}|{executionId}"];
+
+        if (_options.ShowActiveTests)
+        {
+            asm.TestNodeResultsState ??= new(_counter++);
+            asm.TestNodeResultsState.AddRunningTestNode(
+                _counter++, testNodeUid, $"{displayName}", CreateStopwatch());
```
Comment (Evangelink) [inline src/Platform/Microsoft.Testing.Platform/OutputDevice/Terminal/TerminalTestReporter.cs]: Nit: we don't need the interpolated string.

```suggestion
                _counter++, testNodeUid, displayName, CreateStopwatch());
```

File: src/Platform/Microsoft.Testing.Platform/OutputDevice/Terminal/TerminalTestReporter.cs [hunk 6]
```diff
@@ -168,15 +170,15 @@ private TestProgressState GetOrAddAssemblyRun(string assembly, string? targetFra
         return _assemblies.GetOrAdd(key, _ =>
         {
             IStopwatch sw = CreateStopwatch();
-            var assemblyRun = new TestProgressState(assembly, targetFramework, architecture, sw);
+            var assemblyRun = new TestProgressState(_counter++, assembly, targetFramework, architecture, sw);
```
Comment (Evangelink) [inline src/Platform/Microsoft.Testing.Platform/OutputDevice/Terminal/TerminalTestReporter.cs]: For all the places where you do `_counter++`, shall we interlocked them instead to ensure no duplicates?

Comment (drognanar) [reply src/Platform/Microsoft.Testing.Platform/OutputDevice/Terminal/TerminalTestReporter.cs]: @nohwnd is the TerminalTestReporter called in parallel?

File: test/UnitTests/Microsoft.Testing.Platform.UnitTests/OutputDevice/Terminal/TerminalTestReporterTests.cs [hunk 7]
```diff
@@ -167,12 +169,121 @@ Oh no!
...
+    internal sealed class StopwatchFactory
+    {
+        private TimeSpan _currentTime = TimeSpan.Zero;
+
+        public void AddTime(TimeSpan time) => _currentTime += time;
+
+        public IStopwatch CreateStopwatch()
+        {
+            TimeSpan createTime = _currentTime;
+            var stopwatch = new Mock<IStopwatch>();
```
Comment (nohwnd) [inline test/UnitTests/Microsoft.Testing.Platform.UnitTests/OutputDevice/Terminal/TerminalTestReporterTests.cs]: ok. but might be just easier to implement it as a class directly?

File: src/Platform/Microsoft.Testing.Platform/OutputDevice/Terminal/TerminalTestReporter.cs (hunk 6 above)
Comment (nohwnd) [reply src/Platform/Microsoft.Testing.Platform/OutputDevice/Terminal/TerminalTestReporter.cs]: not here but it is in dotnet test. So interlocked would be better.

**Review by nohwnd — APPROVED** [review]

**Review by Evangelink — APPROVED** [review]

**Pull request merged**